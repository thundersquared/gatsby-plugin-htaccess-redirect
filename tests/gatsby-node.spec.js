const {exists, remove, readFile} = require('fs-extra');
const {getRewriteRule, onPostBuild} = require('../gatsby-node');

describe('getRewriteRule', () => {
    it('wraps path in forward slashes', () => {
        expect(getRewriteRule('', 'toPath')).toMatchSnapshot();
    });

    it('allows existing leading and trailing forward slashes', () => {
        expect(getRewriteRule('', '/toPath/')).toMatchSnapshot();
    });

    it('trims leading and trailing whitespace', () => {
        expect(getRewriteRule('', ' toPath ')).toMatchSnapshot();
    });

    it('handles deep paths', () => {
        expect(getRewriteRule('', 'a/b/c/d')).toMatchSnapshot();
    });

    it('handles offset wrapping forward slashes', () => {
        expect(getRewriteRule('', 'a/b/c/')).toMatchSnapshot();
    });

    it('replaces duplicate slashes with single slash', () => {
        expect(getRewriteRule('', 'topath//a')).toMatchSnapshot();
    });

    it('leaves full urls untouched', () => {
        expect(getRewriteRule('', 'http://example.com')).toMatchSnapshot();
        expect(getRewriteRule('', 'http://example.com/')).toMatchSnapshot();
        expect(getRewriteRule('', 'http://example.com/a/b/c')).toMatchSnapshot();
    });

    it('handles redirecting to root', () => {
        expect(getRewriteRule('', '/')).toMatchSnapshot();
    });

    it('handles redirecting to a file', () => {
        expect(getRewriteRule('', '/test.txt')).toMatchSnapshot();
    });

    it('handles redirecting to a file in a folder', () => {
        expect(getRewriteRule('', 'a/b/test.txt')).toMatchSnapshot();
    });
});

describe('onPostBuild', () => {
    const tempFolderPath = './public';

    const assertRedirectFile = async (redirects) => {
        await onPostBuild({
            store: {
                getState: () => ({
                    redirects,
                    program: {
                        directory: './'
                    }
                })
            }
        });

        expect(await exists(`${tempFolderPath}/.htaccess`)).toBe(true);
        expect(await readFile(`${tempFolderPath}/.htaccess`, 'utf-8')).toMatchSnapshot();
    };

    beforeEach(async () => {
        await remove(tempFolderPath);
    });

    // cleanup
    afterAll(async () => {
        await remove(tempFolderPath);
    });

    it('writes redirects from root', async () => {
        await assertRedirectFile(
            [
                {
                    fromPath: '/',
                    toPath: '/hello'
                }
            ],
        );
    });

    it('writes redirects to root', async () => {
        await assertRedirectFile(
            [
                {
                    fromPath: '/hello',
                    toPath: '/'
                }
            ],
        );
    });

    it('writes deep path redirects', async () => {
        await assertRedirectFile(
            [
                {
                    fromPath: '/a/b/c/d',
                    toPath: '/x/y/z'
                }
            ],
        );
    });

    it('handles external redirects', async () => {
        await assertRedirectFile(
            [
                {
                    fromPath: '/a/b',
                    toPath: 'http://example.com/'
                }
            ],
        );
    });
});
