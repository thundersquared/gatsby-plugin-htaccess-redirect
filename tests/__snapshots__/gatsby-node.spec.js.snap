// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getRewriteRule allows existing leading and trailing forward slashes 1`] = `"RewriteRule ^/?/?$ /toPath/ [R=301,L]"`;

exports[`getRewriteRule handles deep paths 1`] = `"RewriteRule ^/?/?$ /a/b/c/d/ [R=301,L]"`;

exports[`getRewriteRule handles offset wrapping forward slashes 1`] = `"RewriteRule ^/?/?$ /a/b/c/ [R=301,L]"`;

exports[`getRewriteRule handles redirecting to a file 1`] = `"RewriteRule ^/?/?$ /test.txt [R=301,L]"`;

exports[`getRewriteRule handles redirecting to a file in a folder 1`] = `"RewriteRule ^/?/?$ /a/b/test.txt [R=301,L]"`;

exports[`getRewriteRule handles redirecting to root 1`] = `"RewriteRule ^/?/?$ / [R=301,L]"`;

exports[`getRewriteRule leaves full urls untouched 1`] = `"RewriteRule ^/?/?$ http://example.com [R=301,L]"`;

exports[`getRewriteRule leaves full urls untouched 2`] = `"RewriteRule ^/?/?$ http://example.com/ [R=301,L]"`;

exports[`getRewriteRule leaves full urls untouched 3`] = `"RewriteRule ^/?/?$ http://example.com/a/b/c [R=301,L]"`;

exports[`getRewriteRule replaces duplicate slashes with single slash 1`] = `"RewriteRule ^/?/?$ /topath/a/ [R=301,L]"`;

exports[`getRewriteRule trims leading and trailing whitespace 1`] = `"RewriteRule ^/?/?$ /toPath/ [R=301,L]"`;

exports[`getRewriteRule wraps path in forward slashes 1`] = `"RewriteRule ^/?/?$ /toPath/ [R=301,L]"`;

exports[`onPostBuild handles external redirects 1`] = `
"<IfModule mod_rewrite.c>
RewriteEngine On
RewriteRule ^/?a/b/?$ http://example.com/ [R=301,L]
</IfModule>"
`;

exports[`onPostBuild writes deep path redirects 1`] = `
"<IfModule mod_rewrite.c>
RewriteEngine On
RewriteRule ^/?a/b/c/d/?$ /x/y/z/ [R=301,L]
</IfModule>"
`;

exports[`onPostBuild writes redirects from root 1`] = `
"<IfModule mod_rewrite.c>
RewriteEngine On
RewriteRule ^/?/?$ /hello/ [R=301,L]
</IfModule>"
`;

exports[`onPostBuild writes redirects to root 1`] = `
"<IfModule mod_rewrite.c>
RewriteEngine On
RewriteRule ^/?hello/?$ / [R=301,L]
</IfModule>"
`;
